src/main/java/org/zalando/nakadi/service/EventStream.java

```diff
@@ -159,13 +183,53 @@ public static String createStreamEvent(final Cursor cursor, final List<String> e
 
     private void sendBatch(final NakadiCursor topicPosition, final List<String> currentBatch)
             throws IOException {
-        // create stream event batch for current partition and send it; if there were
-        // no events, it will be just a keep-alive
-        final String streamEvent = createStreamEvent(cursorConverter.convert(topicPosition), currentBatch);
-        final byte[] batchBytes = streamEvent.getBytes(UTF8);
-        outputStream.write(batchBytes);
-        bytesFlushedMeter.mark(batchBytes.length);
-        outputStream.flush();
+        if(featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.SEND_BATCH_VIA_OUTPUT_STREAM)) {
+            writeStreamEvent(outputStream, cursorConverter.convert(topicPosition), currentBatch);
+        } else {
+            // create stream event batch for current partition and send it; if there were
+            // no events, it will be just a keep-alive
+            final String streamEvent = createStreamEvent(cursorConverter.convert(topicPosition), currentBatch);
+            final byte[] batchBytes = streamEvent.getBytes(UTF8);
+            outputStream.write(batchBytes);
+            bytesFlushedMeter.mark(batchBytes.length);
+            outputStream.flush();
+        }
+    }
+
+    @VisibleForTesting
+    void writeStreamEvent(final OutputStream os, final Cursor cursor, final List<String> events)
+        throws IOException {
+
+        int byteCount = B_FIXED_BYTE_COUNT;
+
+        os.write(B_CURSOR_PARTITION_BEGIN);
+        final byte[] partition = cursor.getPartition().getBytes(StandardCharsets.UTF_8);
+        os.write(partition);
+        byteCount += partition.length;
+        os.write(B_OFFSET_BEGIN);
+        final byte[] offset = cursor.getOffset().getBytes(StandardCharsets.UTF_8);
+        os.write(offset);
+        byteCount += offset.length;
+        os.write(B_CURSOR_PARTITION_END);
+        if (!events.isEmpty()) {
+            os.write(B_EVENTS_ARRAY_BEGIN);
+            for (int i = 0; i < events.size(); i++) {
+                final byte[] event = events.get(i).getBytes(StandardCharsets.UTF_8);
```

Comment:
To get rid of these byte array allocations, it would help to use `java.io.OutputStreamWriter`. Using `java.io.OutputStreamWriter` is slightly more efficient that creating byte arrays from Strings and writing those separately.
Using `com.google.common.io.CountingOutputStream` would simplify the byte counting code.